client/src/app/routes/chrysalis/Addr.tsx

```diff
@@ -357,21 +358,25 @@ class Addr extends AsyncComponent<RouteComponentProps<AddrRouteProps>, AddrState
         this.setState({ transactionHistory: transactionsDetails },
             async () => {
                 const firstPageIndex = (this.state.currentPage - 1) * this.state.pageSize;
-                const lastPageIndex = firstPageIndex + this.state.pageSize;
+                const lastPageIndex = (this.state.currentPage === Math.ceil(this.txsHistory.length / this.state.pageSize))
+                    ? this.txsHistory.length : firstPageIndex + this.state.pageSize;
                 this.updateTransactionHistoryDetails(firstPageIndex, lastPageIndex)
                     .catch(err => console.error(err));
 
-                const allTransactionsDetails = await this._tangleCacheService.transactionsDetails({
-                    network: this.props.match.params.network,
-                    address: this.state.address?.address ?? "",
-                    query: { page_size: Addr.MAX_PAGE_SIZE, state: transactionsDetails?.transactionHistory.state }
-                }, true);
-
-                if (allTransactionsDetails?.transactionHistory.transactions) {
-                    this.setState({ transactionHistory: { ...allTransactionsDetails,
-                        transactionHistory: { ...allTransactionsDetails.transactionHistory,
-                            transactions: [...this.txsHistory, ...allTransactionsDetails.transactionHistory.transactions],
-                            state: allTransactionsDetails.transactionHistory.state } } });
+                if (transactionsDetails?.transactionHistory?.state) {
+                    const allTransactionsDetails = await this._tangleCacheService.transactionsDetails({
+                        network: this.props.match.params.network,
+                        address: this.state.address?.address ?? "",
+                        query: { page_size: Addr.MAX_PAGE_SIZE, state: transactionsDetails?.transactionHistory.state }
+                    }, true);
+
+                    if (allTransactionsDetails?.transactionHistory.transactions) {
+                        this.setState({ transactionHistory: { ...allTransactionsDetails,
+                            transactionHistory: { ...allTransactionsDetails.transactionHistory,
+                                transactions: allTransactionsDetails.transactionHistory
+                                    .transactions?.map(t1 => ({ ...t1, ...this.txsHistory.find(t2 => t2.messageId === t1.messageId) })),
```

Comment:
Should we add an extra `filter` here to make sure we don't end up having `undefined` values or would the `find` always yield the correct tx?

Reply:
Find will always yield correct tx.